views-core/src/main/java/io/micronaut/views/http/RawModelAndViewMessageBodyHandler.java

```diff
@@ -0,0 +1,95 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.views.http;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.io.buffer.ByteBuffer;
+import io.micronaut.core.type.Argument;
+import io.micronaut.core.type.Headers;
+import io.micronaut.core.type.MutableHeaders;
+import io.micronaut.http.HttpHeaders;
+import io.micronaut.http.HttpRequest;
+import io.micronaut.http.MediaType;
+import io.micronaut.http.body.RawMessageBodyHandler;
+import io.micronaut.http.codec.CodecException;
+import io.micronaut.http.context.ServerRequestContext;
+import io.micronaut.views.ModelAndView;
+import io.micronaut.views.ModelAndViewRenderer;
+import io.micronaut.views.exceptions.ViewRenderingException;
+import jakarta.inject.Singleton;
+import org.reactivestreams.Publisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Collection;
+import java.util.Collections;
+
+/**
+ * A message body handler for {@link ModelAndView}.
+ *
+ * @param <T> The model type
+ * @author Tim Yates
+ * @since 6.0.0
+ */
+@Internal
+@Singleton
+final class RawModelAndViewMessageBodyHandler<T> implements RawMessageBodyHandler<ModelAndView<T>> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(RawModelAndViewMessageBodyHandler.class);
+
+    private final ModelAndViewRenderer<T, HttpRequest<?>> modelAndViewRenderer;
+
+    public RawModelAndViewMessageBodyHandler(ModelAndViewRenderer<T, HttpRequest<?>> modelAndViewRenderer) {
+        this.modelAndViewRenderer = modelAndViewRenderer;
+    }
+
+    @Override
+    public @NonNull Collection<? extends Class<?>> getTypes() {
+        return Collections.singletonList(ModelAndView.class);
+    }
+
+    @Override
+    public @NonNull Publisher<ModelAndView<T>> readChunked(@NonNull Argument<ModelAndView<T>> type, @Nullable MediaType mediaType, @NonNull Headers httpHeaders, @NonNull Publisher<ByteBuffer<?>> input) {
+        throw new UnsupportedOperationException("Not supported");
+    }
+
+    @Override
+    public @Nullable ModelAndView<T> read(@NonNull Argument<ModelAndView<T>> type, @Nullable MediaType mediaType, @NonNull Headers httpHeaders, @NonNull InputStream inputStream) throws CodecException {
+        throw new UnsupportedOperationException("Not supported");
+    }
+
+    @Override
+    public void writeTo(@NonNull Argument<ModelAndView<T>> type, @NonNull MediaType mediaType, ModelAndView<T> object, @NonNull MutableHeaders outgoingHeaders, @NonNull OutputStream outputStream) throws CodecException {
+        // The request is required at the moment for the Soy and Pebble renderers; Soy needs to get an Attribute from it, and Pebble needs it to resolve the locale
```

Comment:
can we pass in the attributes and locale without needing the whole request?

Reply:
We need to to pass the whole request as the (HttpLocaleResolver](https://docs.micronaut.io/latest/api/io/micronaut/http/server/util/locale/HttpLocaleResolver.html) takes a whole request. It allows users to resolve the locale from the HTTP Headers, a Query value param or a path of the request.